framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -340,6 +340,29 @@ public final Void visitClass(ClassTree classTree, Void p) {
         try {
             processClassTree(classTree);
             atypeFactory.postProcessClassTree(classTree);
+
+            // If "@B class Y extends @A X {}", then enforce that @B must be a subtype of @A.
+            // classTree.getExtendsClause() is null when there is no explicitly-written extends
+            // clause, as in "class X {}". We assume that this is equivalent to writing
+            // "class X extends @Top Object {}", and there is no need to do any subtype checking.
+            if (classTree.getExtendsClause() != null) {
```

Comment:
I think this logic should go into org.checkerframework.common.basetype.BaseTypeVisitor.processClassTree(ClassTree) which already validates extends and implements clauses.
This code should also validate implements clauses (without duplicating the check).